doc/decisions/DR024-Versioning-traits-and-specifications-method.md

```diff
@@ -0,0 +1,365 @@
+    if instanceOf(batchElementError, VersionNotSupportedError):
+      incompatible_with_manager_error()
+    ...
+
+self._manager.resolve(
+    [entity_reference], trait_set, ResolveAccess.kRead, context, success_cb, error_cb
+)
+```
+### Manager side
+```python
+# On the manager side, each trait gets unpacked (probably via a utility
+# method provided by traitgen or openassetio), and branching is done
+# on a trait by trait basis
+
+def resolve(
+        self,
+        entityReferences,
+        traitSet,
+        resolveAccess,
+        context,
+        hostSession,
+        successCallback,
+        errorCallback,
+    ):
+
+  # This function is imagined a pre-provided utility, possibly generated
+  # via traitgen, that would error in some way if the trait set does not
+  # contain the locatableContent trait.
+  locatableContent_version = locatableContent_version_from_trait_set(traitSet)
+
+  success_result = TraitsData()
+  if (locatableContent_version == 2):
+    trait = mediacreation.traits.content.LocatableContentTrait(success_result)
+    trait.setUrl(managed_filesystem_locations[ref.toString()])
+    successCallback(idx, success_result)
+  elif (locatableContent_version == 1):
+    trait = mediacreation.legacy.traits.content.LocatableContentTrait_v1(success_result)
+    trait.setLocation(managed_filesystem_locations[ref.toString()])
+    successCallback(idx, success_result)
+  else:
+    raise RuntimeError()
+
+```
+
+#### Pros
+- Both managers and hosts have the option to be backwards compatible.
+- More flexible, traits of different versions can be mixed in the same
+  session, which could allow hosts/managers to perform partial upgrades
+  in the case they are blocked on small specifics.
+- Hosts don't have to interrogate and declare which library versions
+  are in use, which helps especially in the case of second-order
+  dependencies.
+- Seems to fit with existing OpenAssetIO design, especially with
+  `BatchElementError`, as an incompatible version wouldn't necessarily
+  need to error the whole batch.
+- No new API, core `openassetio` library remains version unaware.
+- Cannot be ignored, unlike other "layered on top" mechanisms.
+
+#### Cons
+- No "primary" version across the session, increasing implementation
+  complexity.
+- Introduces "Try and see" negotiation on the host side, although
+  introspection behaviour to negate this could be introduced at a later
+  date, similar to other introspection mechanisms. (See Option 3 -
+  Extended)
+- Would likely require some utilities in `traitgen`, otherwise, due to
+  each new version effectively being a brand new trait, constructing the
+  correct trait view would become a tedious dance.
+- EntityTraits becomes more complex, as we either have to declare it
+  version agnostic somehow, or managers would need to return the correct
+  versions/entire swathes of versions.
+- Unclear how specification versioning is to work without an out-of-data
+  version.
```

Comment:
Isn't this a bit too strong? Specifications will still "work", they'd just be unversioned, and unable to be versioned with a direct in-data representation. I grant the significant rework think is probably worth saying.

Reply:
It might be good to clarify what you mean by "un versioned". 

The existing specifications approach would inherently version the traits - without change. What you wouldn't get with the current design was the ability to version the specifications themselves in a way that was visible in the resulting data, which is certainly a con against pre-negotiation of a package specific version.

At present however, Specifications deliberately don't exist at the data/core API level. They're just shorthand for a particular trait set.

Adding the ability to store alternate trait sets as different versions of a specification in the generated view classes doesn't really seem any more onerous that generating multiple classes for the trait views themselves. 

If you want to go from a trait set to a spec view class, it doesn't seem any different from as it stands without versioning- in that there is no concrete 1:1 mapping anyway. There just may be n number of well-known specifications that overlap a given trait set. There are just more known view classes if you have versioned specs. That problem exists beyond the scope of versioning I think as it will always run up against one traitset being a subset of another (which is deliberate, by design).

The main downside seems to be about having hosts/managers agree on _specifications_ as an atomic unit instead of a trait set, which could be possible with package versioning and the out-of-data mechanism.

I think perhaps a fairer version of this is that with in-data versioning, specifications work exactly as they do now, but out-of-data has a pro, that it would be possible to have more consensus about which specifications to use directly.

Reply:
> At present however, Specifications deliberately don't exist at the data/core API level. They're just shorthand for a particular trait set.

Ruminating on this a little - wondered if it might be worth just highlighting why this is currently the case. As it might suggest whether knowing specification versions is actually important or not beyond programmer convenience.

The reason they’re not expressed at the API level is that the interoperability concept of the API relies specifically on specific sets of _traits_ not the specification that happened to make some data, to make programmers lives easier. In the majority of cases one side of the equation only probably cares about a subset of these traits in particular - and this is what allows interop.

For example, the read node in nuke only cares that entity has locatable content. In the future, it may care about image related traits. As long as target entity has those traits as a _subset_ of its traits then all is well. Nuke may use an ImageSpecification of sorts for consistency, but it doesnt matter if the original entity was authored by some DelightRenderAOVSpecificstion view class or by hand in a test case json file, or made up on the fly, as long as it has compatible versions of the specific traits - locatable content etc… 

There is the mirror in publishing where a manager may only want to know if it has locatable content to store and just stash the rest for filtering later.

There are certainly cases where specific combinations may be required, and versioning the classes used can help with alignment, but ultimately it’s just the traits themselves that actually make any meaningful difference - which is why specifications don’t exist at the API level today. All a spec is doing is helping check which traits are there.

Something @elliotcmorris often mentioned is - why do we even need them? I imagine we probably don’t, compatibility of data is implicit in the trait set so doesn’t need another layer, they’re just there to help communicate which traits are mutually agreed on by some subset of a user base when alignment is helpful.

Reply:
> Adding the ability to store alternate trait sets as different versions of a specification in the generated view classes doesn't really seem any more onerous that generating multiple classes for the trait views themselves.

I agree on this, this would be the "significant rework", but I don't think it's enormous. it's more of a consistency problem, and entirely within the domain of programmer convenience. Given an in-data approach, how does a method know which specific version of a specification it should be expecting to work with, you're left with a very nuts and bolts try-and-see approach, is a host expected to scan through managementPolicy with every possible specification version until it gets the one the manager happens to support? Seem onerous, and is not something that would be expected currently.

>  think perhaps a fairer version of this is that with in-data versioning, specifications work exactly as they do now, 

Whilst true, surely specifications must eventually be versioned somehow if they are to remain useful. If anything, I would expect the structure of specifications to be the most volatile over time, if specifications end up being used as the first class units of exchange they are intended to be. The downside of in-data versioning is that as a strategy, it will not work on specifications precisely because they are not concepts that exist in the api. This strikes me as a downside, as we were expecting an hoping for a solution that worked for both traits & specs (which you can understand when you view this exercise from the perspective of adding versions to the items defined in the .yml file)

> ultimately it’s just the traits themselves that actually make any meaningful difference - which is why specifications don’t exist at the API level today. All a spec is doing is helping check which traits are there.

Something @elliotcmorris often mentioned is - why do we even need them?

I continue to mention this, to the disdain of my colleagues. It's something i've never quite been able to take on faith, the usefulness of these trait groupings, as from what I've seen, and what I expect, managers and hosts will _mostly_ treat traits as atomic units, and perhaps have some exceptions for certain groupings (traitsets,) but i'm not sure it'll be the standard practice, so I continue to question the usefullness of specifications in general, and wonder if their inability to be versioned should stand in the way here.

Reply:
> Given an in-data approach, how does a method know which specific version of a specification it should be expecting to work with,

That's what I was trying to get at in the second reply - it doesn't matter what specification is used, as long as the traits match. There never will be a 1:1 mapping of a trait set to a specification, due to the whole "add more traits to be more specific" thing. The traits are the only important thing.

>  perhaps have some exceptions for certain groupings (traitsets,)

To me thats all they're for - they're a "light weight convenience" not a core-currency, so I totally agree with you, that if it makes this versioning thing easier to grok, just ignore them! I really don't believe they belong at the API level, just as you said - to help communicate expectations for certain groupings.

Reply:
I'm going to leave this as a con, as leaving things as they are, I believe even just to a new user looking at mediacreation, specification versioning will become a question we'll have to be able to explain.